sqlalchemy-stubs/sql/elements.pyi

```diff
@@ -47,300 +61,388 @@ class ColumnElement(operators.ColumnOperators, ClauseElement):
 
-class FunctionFilter(ColumnElement):
+class FunctionFilter(ColumnElement[_TE]):
     __visit_name__: str = ...
     criterion: Any = ...
-    func: Any = ...
-    def __init__(self, func, *criterion) -> None: ...
-    def filter(self, *criterion): ...
-    def over(self, partition_by: Optional[Any] = ..., order_by: Optional[Any] = ...): ...
+    func: FunctionElement[_TE] = ...
+    def __init__(self, func: FunctionElement[_TE], *criterion: Any) -> None: ...
+    def filter(self, *criterion: Any) -> FunctionFilter[_TE]: ...
+    def over(self, partition_by: Optional[Union[str, ColumnElement[Any], Iterable[Union[str, ColumnElement[Any]]]]] = ...,
+             order_by: Optional[Union[str, ColumnElement[Any], Iterable[Union[str, ColumnElement[Any]]]]] = ...) -> Over[_TE]: ...
     @property
-    def type(self): ...
-    def get_children(self, **kwargs): ...
+    def type(self) -> _TE: ...
+    def get_children(self, **kwargs: Any) -> List[Any]: ...
 
-class Label(ColumnElement):
+_L = TypeVar('_L', bound=Label)
+
+class Label(ColumnElement[_TE]):
     __visit_name__: str = ...
-    name: Any = ...
-    key: Any = ...
-    def __init__(self, name, element, type_: Optional[Any] = ...) -> None: ...
-    def __reduce__(self): ...
+    name: str = ...
+    key: str = ...
+    @overload
+    def __init__(self, name: str, element: ColumnElement[_TE]) -> None: ...
+    @overload
+    def __init__(self, name: str, element: ColumnElement[Any], type_: Type[_TE] = ...) -> None: ...
+    @overload
+    def __init__(self, name: str, element: ColumnElement[Any], type_: _TE = ...) -> None: ...
+    def __reduce__(self) -> Any: ...
     @property
-    def type(self): ...
+    def type(self) -> Union[_TE, NullType]: ...  # type: ignore  # This return type is not compatible with ColumnElement.type
     @property
-    def element(self): ...
-    def self_group(self, against: Optional[Any] = ...): ...
+    def element(self) -> ColumnElement[_TE]: ...
+    def self_group(self, against: Optional[Any] = ...) -> Union[ColumnElement[_TE], Grouping[_TE]]: ...
     @property
-    def primary_key(self): ...
+    def primary_key(self) -> Any: ...
     @property
-    def foreign_keys(self): ...
-    def get_children(self, **kwargs): ...
+    def foreign_keys(self) -> List[Any]: ...  # type: ignore
+    def get_children(self, **kwargs: Any) -> Any: ...
 
-class ColumnClause(Immutable, ColumnElement):
+class ColumnClause(Immutable, ColumnElement[_TE]):
     __visit_name__: str = ...
     onupdate: Any = ...
     default: Any = ...
     server_default: Any = ...
     server_onupdate: Any = ...
-    key: Any = ...
-    table: Any = ...
-    type: Any = ...
-    is_literal: Any = ...
-    def __init__(self, text, type_: Optional[Any] = ..., is_literal: bool = ...,
-                 _selectable: Optional[Any] = ...) -> None: ...
+    key: str = ...
+    name: str = ...
+    table: Optional[TableClause] = ...
+    type: Union[_TE, NullType] = ...  # type: ignore
```

Comment:
Adding `NullType` here (and maybe also elsewhere) is probably a bad idea. Many columns are defined as `nullable=False` and this causes false positives. Later we can write a plugin (or use literal types when they will be added to give more precise type). But currently we should use types that generate as few false positives as possible.

Reply:
I agree. This would be simpler if we could determine `type` based on which `__init__` overload was chosen, but I don't think we can at this time.